scripts/binary/smoke.js

```diff
@@ -201,7 +201,96 @@ const runV8SnapshotProjectTest = function (buildAppExecutable, e2e) {
+      `--spec=${e2e}/cypress/e2e/simple_passing.cy.js`,
+    ]
+
+    if (verify.needsSandbox()) {
+      args.push('--no-sandbox')
+    }
+
+    const options = {
+      stdio: ['inherit', 'inherit', 'pipe'], env,
+    }
+
+    console.log('running project test')
+    console.log(buildAppExecutable, args.join(' '))
+
+    const childProcess = cp.spawn(buildAppExecutable, args, options)
+    let errorOutput = ''
+
+    childProcess.stderr.on('data', (data) => {
+      errorOutput += data.toString()
+    })
+
+    childProcess.on('exit', (code) => {
+      if (code === 0) {
+        return reject(new Error(`running project tests should have failed for test: ${testName}`))
+      }
+
+      if (!errorOutput.includes(errorMessage)) {
+        return reject(new Error(`running project tests failed with errors: ${errorOutput} but did not include the expected error message: '${errorMessage}'`))
+      }
+
+      return resolve()
+    })
+  })
+}
+
+const runIntegrityTest = async function (buildAppExecutable, buildAppDir, e2e) {
+  const testCorruptingFile = async (file, errorMessage) => {
+    const contents = await fs.readFile(file)
+
+    // Backup state
+    await fs.move(file, `${file}.bak`)
+
+    // Modify app
+    await fs.writeFile(file, Buffer.concat([contents, Buffer.from(`\nconsole.log('modified code')`)]))
+    await runErroringProjectTest(buildAppExecutable, e2e, `corrupting ${file}`, errorMessage)
+
+    // Restore original state
+    await fs.move(`${file}.bak`, file, { overwrite: true })
+  }
+
+  await testCorruptingFile(path.join(buildAppDir, 'index.js'), 'Error: Integrity check failed for main index.js file')
+  await testCorruptingFile(path.join(buildAppDir, 'packages', 'server', 'index.jsc'), 'Error: Integrity check failed for main server index.jsc file')
+  await testCorruptingFile(path.join(buildAppDir, 'node_modules', 'bytenode', 'lib', 'index.js'), 'Error: Integrity check failed for main bytenode.js file')
+
+  const testAlteringEntryPoint = async (additionalCode, errorMessage) => {
+    const packageJsonContents = await fs.readJSON(path.join(buildAppDir, 'package.json'))
+
+    // Backup state
+    await fs.move(path.join(buildAppDir, 'package.json'), path.join(buildAppDir, 'package.json.bak'))
+
+    // Modify app
+    await fs.writeJSON(path.join(buildAppDir, 'package.json'), {
+      ...packageJsonContents,
+      main: 'index2.js',
+    })
+
+    await fs.writeFile(path.join(buildAppDir, 'index2.js'), `${additionalCode}\nrequire("./index.js")`)
+    await runErroringProjectTest(buildAppExecutable, e2e, 'altering entry point', errorMessage)
+
+    // Restore original state
+    await fs.move(path.join(buildAppDir, 'package.json.bak'), path.join(buildAppDir, 'package.json'), { overwrite: true })
+    await fs.remove(path.join(buildAppDir, 'index2.js'))
+  }
+
+  await testAlteringEntryPoint('console.log("simple alteration")', 'Error: Integrity check failed with expected stack length 9 but got 10')
```

Comment:
Is `Error: Integrity check failed with expected stack length 9 but got 10` the exact error the user will see? I feel like we could add some extra information/guidance, something like "We detected an issue with the integrity of the Cypress binary. It may have been compromised and is not safe to run. We recommend re-installing the Cypress binary with `cypress cache clear && cypress install`."

Reply:
That's a good call. I can make them a little more user friendly.

Reply:
Updated: [`5f3a735` (#24909)](https://github.com/cypress-io/cypress/pull/24909/commits/5f3a7351474e241ae20cac5ec488f7bd8a146c19)